### PR TITLE
The higgler

### PR DESCRIPTION
**Pull request by isovector**
@i-am-tom we've gone too far

### CODE REVIEW

File: example/Main.hs [hunk 1]
```diff
@@ -15,18 +15,14 @@ import Weft.Generics.Hydrate
...
-
 data GqlQuery ts = GqlQuery
     { getUser :: Magic ts (Arg "id" ID -> User ts)
     , getAllUsers :: Magic ts [User ts]
     } deriving (Generic)
 
 data User ts = User
   { userId         :: Magic ts (Arg "arg" (Maybe Int) -> ID)
-  , userName       :: Magic ts Name
+  , userName       :: Magic ts String
```
Comment (JonathanLorimer) [inline example/Main.hs:23]: could try and add Name back

File: src/Weft/Generics/EmptyQuery.hs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 module Weft.Generics.EmptyQuery
```
Comment (JonathanLorimer) [inline src/Weft/Generics/EmptyQuery.hs]: remove empty query

File: src/Weft/Generics/PprQuery.hs [hunk 3]
```diff
@@ -1,11 +1,15 @@
+{-# LANGUAGE CPP          #-}
```
Comment (JonathanLorimer) [inline src/Weft/Generics/PprQuery.hs]: don't need this

File: src/Weft/Generics/QueryParser.hs [hunk 4]
```diff
@@ -54,9 +65,55 @@ anonymousQueryParser = do
 class GPermFieldsParser (rq :: * -> *) where
   gPermFieldsParser :: [ReaderT Vars Parser (rq x)]
 
-instance {-# OVERLAPPABLE #-} GPermFieldsParser fq => GPermFieldsParser (M1 a b fq) where
+instance  GPermFieldsParser fq => GPermFieldsParser (M1 D b fq) where
```
Comment (JonathanLorimer) [inline src/Weft/Generics/QueryParser.hs]: this can just be one generic case

File: src/Weft/Generics/RecordGen.hs [hunk 5]
```diff
@@ -31,27 +42,66 @@ instance (GRecordGen r1, GRecordGen r2) => GRecordGen (r1 :*: r2) where
...
+  termGen = ToMagic <$> termGen
+
+instance GRecordGen (M1 _1 _2 _3) => TermGen (M1 _1 _2 _3 Void) where
+  termGen = gRecordGen
+
+instance TermGen a => TermGen [a] where
+  termGen = sized $ \n ->
+    case n <= 0 of
+      True  -> pure []
+      False -> resize (max 0 $ n - 1) $ listOf termGen
```
Comment (JonathanLorimer) [inline src/Weft/Generics/RecordGen.hs]: max isn't necessary

File: src/Weft/Generics/Schema.hs [hunk 6]
```diff
@@ -29,20 +39,42 @@ schema
...
 
 instance {-# OVERLAPPING #-} (KnownSymbol a, HasGqlType t, ReifyArgs args)
-      => GHasSchema (M1 S ('MetaSel ('Just a) b c d) (Rec0 t))
-                    (M1 S ('MetaSel ('Just a) b c d) (Rec0 (Field args))) where
+      => GHasSchema (M1 S ('MetaSel ('Just a) b c d) (K1 _1 t))
+                    (M1 S ('MetaSel ('Just a) b c d) (K1 _1 (Field args))) where
     gSchema = M1 $ K1 $ Field (reifyNameType @a @t) (reifyArgs @args)
 
+
+#define INST(magic, ts) (M1 S ('MetaSel ('Just a) b c d) (K1 _1 (magic 'ts t)))
```
Comment (JonathanLorimer) [inline src/Weft/Generics/Schema.hs:64]: normalize

File: test/ParserSpec.hs [hunk 7]
```diff
@@ -5,13 +5,15 @@ module ParserSpec where
 import           Control.Monad.Reader
 import           Data.Bifunctor
 import           Data.Either
+import qualified Data.List as L
 import qualified Data.Map as M
 import           Data.Text (Text)
 import qualified Data.Text as T
-import qualified Data.List as L
+import           Data.Void
```
Comment (JonathanLorimer) [inline test/ParserSpec.hs:12]: move this to types

**Review by JonathanLorimer — COMMENTED** [review]
Looks good, left some comments

File: example/Main.hs (hunk 1 above)
Comment (isovector) [reply example/Main.hs:23]: don't work! not sure why -- will investigate later

**Comment by i-am-tom** [comment]
Madre de dios

**Pull request merged**